### PR TITLE
Add words about read refreshing to the transaction design page

### PR DESCRIPTION
**Pull request by andreimatei**

### CODE REVIEW

**Comment by andreimatei** [comment]
cc @jordanlewis who requested some docs for the refreshing

File: v2.0/architecture/transaction-layer.md [hunk 1]
```diff
@@ -132,27 +126,32 @@ CockroachDB efficiently supports the strongest ANSI transaction isolation level:
...
 
 2. `TxnB` tries to push `TxnA`'s timestamp forward.
 
     This succeeds only in the case that `TxnA` has snapshot isolation and `TxnB`'s operation is a read. In this case, the [write skew](https://en.wikipedia.org/wiki/Snapshot_isolation) anomaly occurs.
 
 3. `TxnB` enters the `PushTxnQueue` to wait for `TxnA` to complete.
 
+Additionally, the following types of conflicts that don't involve running into intents can arise:
+
+- **Write after read**, when a write with a lower timestamp encouters a later read. This is handled through the [Timestamp Cache](#timestamp-cache).
```
Comment (spencerkimball) [inline v2.0/architecture/transaction-layer.md]: s/encouters/encounters/

File: v2.0/architecture/transaction-layer.md [hunk 2]
```diff
@@ -132,27 +126,32 @@ CockroachDB efficiently supports the strongest ANSI transaction isolation level:
...
+- **Read within uncertainty window**, when a read encounters a value with a higher timestamp but it's ambiguous whether the value should be considered to be in the future or in the past of the transaction because of possible *clock skew*. This is handled by attempting to push the transaction's timestamp beyond the uncertain value (see [read refreshing](#read-refreshing)).
```
Comment (spencerkimball) [inline v2.0/architecture/transaction-layer.md]: s/attempting to push.../attempting to forward the transaction's timestamp just beyond the uncertain value's timestamp/ (I don't think push is right in this context).

You might also add: "If the transaction timestamp forwarding and refreshing is unsuccessful, then the transaction must be retried at the forwarded timestamp. Note that on successive retries, reads will never encounter uncertainty issues on any node which was previously visited."

File: v2.0/architecture/transaction-layer.md [hunk 3]
```diff
@@ -172,6 +171,16 @@ Blocked transactions also check the status of their own transaction to ensure th
...
+
+Whenever a write occurs, its timestamp is checked against the Timestamp Cache. If the timestamp is less than the Timestamp Cache's latest value, we attempt to push the timestamp for its transaction forward to a later time. In the case of serializable transactions, this might cause them to restart in the second phase of the transaction (see [read refreshing](#read-refreshing)).
+
+### Read Refreshing
+
+Whenever a transaction's timestamp has been pushed, additional checks are required before allowing serializable transactions to commit at the pushed timestamp: we need to check that whatever the transaction has previously read at its original timestamp hasn't changed before the pushed timestamp. This is done by keeping track of all the reads using a dedicted `RefreshRequest`. If this succeeds, the transaction is allowed to commit (technically, transactions perform this check at commit time if they've been pushed by a different transaction or by the timestamp cache, or they perform the check whenever they encounter a `ReadWithinUncertaintyIntervalError` before continuing).
```
Comment (spencerkimball) [inline v2.0/architecture/transaction-layer.md]: I would re-word this to remove the use of first person pronouns ("we"). How about:

"any values which the transaction previously read must be checked to verify that no writes have subsequently occurred between the original transaction timestamp and the forwarded transaction timestamp. This check prevents serializable isolation violations."

s/dedicted/dedicated/

s/technically, transactions/usually transactions/
s/before continuing/immediately, before continuing/

**Comment by jordanlewis** [comment]
@andreimatei thanks for embarking on this effort.

**Comment by andreimatei** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[v2.0/architecture/transaction-layer.md, line 152 at r1](https://beta.reviewable.io/reviews/cockroachdb/docs/2684#-L76pQ1U4CqokbAMth7L-r1-152:-L7BQ4I4DR7V4yEwnvMm:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b675afe0b70347f0ff381f3a21b1f89cb7b751de/v2.0/architecture/transaction-layer.md#L152)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

s/encouters/encounters/
</blockquote></details>

Done.

---

*[v2.0/architecture/transaction-layer.md, line 153 at r1](https://beta.reviewable.io/reviews/cockroachdb/docs/2684#-L76pQ1U4CqokbAMth7L-r1-153:-L7BQHvfBMM0WqdlPwl-:b2isg5g) ([raw file](https://github.com/cockroachdb/docs/blob/b675afe0b70347f0ff381f3a21b1f89cb7b751de/v2.0/architecture/transaction-layer.md#L153)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

s/attempting to push.../attempting to forward the transaction's timestamp just beyond the uncertain value's timestamp/ (I don't think push is right in this context).

You might also add: "If the transaction timestamp forwarding and refreshing is unsuccessful, then the transaction must be retried at the forwarded timestamp. Note that on successive retries, reads will never encounter uncertainty issues on any node which was previously visited."
</blockquote></details>

why isn't "push" the right word? That's what we generally use when we talk about a transaction's timestamp being changed. I understand what we discussed the other day - about how, in the case of writes running into the timestamp cache the transaction can continue reading at its old timestamp, but in the case of uncertainty errors, the txn needs to immediately start reading at the new timestamp, but I don't think these considerations make "push" not valid...

I've incorporated your addendum, but I've split it between here and the refresh section.

---

*[v2.0/architecture/transaction-layer.md, line 182 at r1](https://beta.reviewable.io/reviews/cockroachdb/docs/2684#-L76pQ1U4CqokbAMth7L-r1-182:-L7BSVhj-XH1Fbg6qL42:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b675afe0b70347f0ff381f3a21b1f89cb7b751de/v2.0/architecture/transaction-layer.md#L182)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

I would re-word this to remove the use of first person pronouns ("we"). How about:

"any values which the transaction previously read must be checked to verify that no writes have subsequently occurred between the original transaction timestamp and the forwarded transaction timestamp. This check prevents serializable isolation violations."

s/dedicted/dedicated/

s/technically, transactions/usually transactions/
s/before continuing/immediately, before continuing/
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/docs/2684)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[v2.0/architecture/transaction-layer.md, line 182 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2684#-L76pQ1U4CqokbAMth7L-r1-182:-L7BjDfvQm7JVv3fdATW:b13zqa3) ([raw file](https://github.com/cockroachdb/docs/blob/b675afe0b70347f0ff381f3a21b1f89cb7b751de/v2.0/architecture/transaction-layer.md#L182)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I thought the reads were tracked at the `TxnCoordSender` (or whatever it is now) and "refreshed" using `RefreshRequest`. The wording above makes it sound like the `RefreshRequest` is doing the tracking. Admittedly, I'm not familiar with the code, so perhaps the above is accurate.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2684)*

**Comment by andreimatei** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[v2.0/architecture/transaction-layer.md, line 182 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2684#-L76pQ1U4CqokbAMth7L-r1-182:-L7BkdUZ2MRyAe1iz8X-:bcuacvj) ([raw file](https://github.com/cockroachdb/docs/blob/b675afe0b70347f0ff381f3a21b1f89cb7b751de/v2.0/architecture/transaction-layer.md#L182)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I thought the reads were tracked at the `TxnCoordSender` (or whatever it is now) and "refreshed" using `RefreshRequest`. The wording above makes it sound like the `RefreshRequest` is doing the tracking. Admittedly, I'm not familiar with the code, so perhaps the above is accurate.
</blockquote></details>

There's a missing word indeed; I wanted to say "keeping track of all the reads *and* using a dedicated `RefreshRequest`"
Will amend, thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2684)*

**Pull request merged**